src/RecurringGrantDrop.sol

```diff
@@ -185,26 +199,96 @@ contract RecurringGrantDrop is Ownable2Step{
+        external
+    {
+        uint256 grantId = grant.calculateId(timestamp);
+        checkClaimReserved(timestamp, receiver, root, nullifierHash, proof, signature);
+
+        nullifierHashes[nullifierHash] = true;
+
+        SafeERC20.safeTransferFrom(token, holder, receiver, grant.getAmount(grantId));
+
+        emit GrantClaimed(grantId, receiver);
+    }
+
+    /// @notice Check whether a claim is valid
+    /// @param timestamp The timestamp of the reservation
+    /// @param receiver The address that will receive the tokens (this is also the signal of the ZKP)
+    /// @param root The root of the Merkle tree (signup-sequencer or world-id-contracts provides this)
+    /// @param nullifierHash The nullifier for this proof, preventing double signaling
+    /// @param proof The zero knowledge proof, array of 8 uint256 elements, demonstrating that the claimer has a verified World ID
+    /// @param signature The off-chain signature of the reservation.
+    function checkClaimReserved(uint256 timestamp, address receiver, uint256 root, uint256 nullifierHash, uint256[8] calldata proof, bytes calldata signature)
+        public
+    {
+        uint256 grantId = grant.calculateId(timestamp);
+
+        if (receiver == address(0)) revert InvalidReceiver();
+        if (timestamp > block.timestamp) revert InvalidTimestamp();
+
+        this.checkNullifier(grantId, receiver, root, nullifierHash, proof);
+
+        grant.checkReservationValidity(timestamp);
+
+        address signer = ECDSA.recover(keccak256(abi.encode(timestamp, nullifierHash)), signature);
+        if (!allowedSigners[signer]) revert UnauthorizedSigner();
+
+        worldIdRouter.verifyProof(
+            root,
+            groupId,
+            uint256(keccak256(abi.encodePacked(receiver))) >> 8,
+            nullifierHash,
+            grantId,
+            proof
+        );
+    }
+
+    /// @notice Check whether a nullifier has been used before in the previous contract.
+    function checkNullifier(uint256 grantId, address receiver, uint256 root, uint256 nullifierHash, uint256[8] calldata proof) external {
+        if (nullifierHashes[nullifierHash]) revert InvalidNullifier();
+
+        // If this is for a grant before the new deployment, check the previous contract.
+        if (grantId < RELEASE_GRANT_ID) {
+            try PREVIOUS_CONTRACT.checkClaim(grantId, receiver, root, nullifierHash, proof) {
+                // This should not happen since only claiming the current grant can succeed, which is already prohibited by the check above.
+            } catch (bytes memory reason) {
+                // Check if nullifier has already been used.
+                if (bytes4(reason) == RecurringGrantDrop.InvalidNullifier.selector) {
+                    revert InvalidNullifier();
+                }
+                // Any other error is fine.
```

Comment:
The problem I have with this is that 'any other error' can be anything, like a revert in the `grantId` check or an overflow or anything. This only works if we are absolutely sure the duplicate nullifier check is the very first revert opportunity, which is imo too strong an assumption.

Reply:
Now that only the `Legacy` contract has this ugly hack and only calls out to the [non-upgradable existing contract](https://optimistic.etherscan.io/address/0xe773335550b63eed23a6e60dcc4709106a1f653c#code), I think it's fine to keep it like this in there, since we know that the assumption will hold.